lib/orderbook/MatchesProcessor.ts

```diff
@@ -1,38 +1,90 @@
+import uuidv1 from 'uuid/v1';
+import { SwapDeal } from './SwapDeals';
+import { SwapDealRole } from '../types/enums';
 
 class MatchesProcessor {
-  private buffer: any[];
-
-  constructor(private logger: Logger, private pool?: Pool, private raidenClient?: RaidenClient) {
-    this.buffer = [];
-  }
-
-  public add(match: OrderMatch) {
-    this.buffer.push(match);
+  constructor(private logger: Logger, private pool?: Pool, private raidenClient?: RaidenClient) { }
 
+  public process = (match: OrderMatch) => {
     if (isPeerOrder(match.maker)) {
-      this.notifyPeer(match);
-    } else if (isPeerOrder(match.taker)) {
       this.executeSwap(match);
+    } else if (isPeerOrder(match.taker)) {
+      this.notifyPeer(match);
     } else {
       this.notifyClient(match);
     }
   }
 
-  public notifyPeer(_args: any) {
+  private notifyPeer(_match: OrderMatch) {
     // TODO: notify the peer to trigger swap execution from his side
   }
 
-  public notifyClient(_args: any) {
+  private notifyClient(_match: OrderMatch) {
     // TODO: notify the local exchange client on the match
   }
 
-  public executeSwap(_args: any) {
-    // TODO: execute the swap procedure
+  private executeSwap = (match: OrderMatch) => {
+    if (this.pool) {
+      /** The remote maker order we are filling. */
+      const maker = match.maker as StampedPeerOrder;
+      /** Our local taker order. */
+      const taker = match.taker as StampedOwnOrder;
+
+      const peer = this.pool.getPeer(maker.peerPubKey);
+
+      const [baseCurrency, quoteCurrency] = maker.pairId.split('/');
+
+      let takerCurrency: string;
+      let makerCurrency: string;
+      let takerAmount: number;
+      let makerAmount: number;
+      if (taker.quantity > 0) {
+        // we are buying the base currency
+        takerCurrency = baseCurrency;
+        makerCurrency = quoteCurrency;
+        takerAmount = taker.quantity;
+        makerAmount = taker.quantity * maker.price;
+      } else {
+        // we are selling the base currency
+        takerCurrency = quoteCurrency;
+        makerCurrency = baseCurrency;
+        takerAmount = taker.quantity * maker.price;
+        makerAmount = taker.quantity;
+      }
+
+      const dealRequestBody: packets.DealRequestPacketBody = {
+        takerCurrency,
+        makerCurrency,
+        takerAmount,
+        makerAmount,
+        takerPubKey: this.pool.handshakeData.nodePubKey,
+        takerDealId: uuidv1(),
+        orderId: maker.id,
+      };
+
+      const deal: SwapDeal = {
+        ...dealRequestBody,
+        myRole : SwapDealRole.Taker,
+        createTime: Date.now(),
+      };
+
+      // TODO: Put swapped amount on hold
+
+      this.pool.swapDeals.add(deal);
+      this.logger.debug('Initiating swap deal: ' + JSON.stringify(deal));
+
+      // TODO: Subscribe to swap events to handle swap resolution
+
+      const packet = new packets.DealRequest(dealRequestBody);
+
+      peer.sendPacket(packet);
+    }
```

Comment:
Was this implemented separately from `Service.executeSwap` because you wanted to postpone dealing with that and with #380 to a later stage? or was there a different reason?

Reply:
Yes that's the main reason, also they just function a bit different because `Service.executeSwap` doesn't pertain to any particular order.